src/lib.rs

```diff
@@ -0,0 +1,277 @@
+pub struct BbsParams<G: Group> {
+    h_generators: Vec<G>,
+    g_generator: G,
+}
+
+impl BbsParams<G1Projective> {
+    /// Generates parameters for the BBS scheme based on the `num_attributes` parameter:
+    ///
+    /// # Arguments
+    /// - `num_attributes`: the number of attributes to be signed
+    /// - `rng`: a random number generator
+    ///
+    /// # Returns
+    /// - `BbsParams`: the system parameters
+    ///
+    /// # Example
+    /// ```rust
+    /// use bbs_mac_vs_sig::BbsParams;
+    /// let mut rng = rand::thread_rng();
+    /// let num_attributes = 5;
+    /// let params = BbsParams::new(num_attributes, &mut rng);
+    /// ```
+    pub fn new(num_attributes: usize, rng: &mut impl RngCore) -> Self {
+        let g_generator = G1Projective::generator();
+        let mut h_generators = Vec::with_capacity(num_attributes);
+        for _ in 0..num_attributes {
+            h_generators.push(G1Projective::random(&mut *rng));
+        }
+        Self {
+            h_generators,
+            g_generator,
+        }
+    }
+}
+
+/// A BBS MAC (Message Authentication Code) with the following components:
+/// - `a`: the MAC value `A = (x + e)^-1 * (G + m1*H1 + ... + ml*Hl)` for `l` messages
+/// - `e`: the random scalar used for the MAC
+#[derive(Debug)]
+pub struct BbsMac {
+    a: G1Projective,
+    e: Scalar,
+}
+
+#[derive(Debug)]
+pub enum BbsMacError {
+    /// thrown when the number of messages does not match the number of generators
+    InputLengthMismatch,
+    /// thrown when the inverse of (x + e) cannot be computed (e.g. when x + e = 0)
+    VerificationInverseFailed,
+    /// thrown when the verification fails (e.g. when the given
+    /// `A` != `(x + e)^-1 * (G + m1*H1 + ... + ml*Hl)`)
+    VerificationFailed,
+}
+
+impl BbsMac {
+    /// Implements the core BBS signing computation:
+    /// A = (x + e)^-1 * (G + m1*H1 + ... + ml*Hl)
+    ///
+    /// # Arguments
+    /// - `bbs_params`: the system parameters
+    /// - `x`: the secret key scalar
+    /// - `messages`: the messages to be signed
+    /// - `rng`: a random number generator
+    ///
+    /// # Returns
+    /// - `Ok(BbsMac)`: the computed BBS MAC
+    /// - `Err(BbsMacError::InputLengthMismatch)`: an error if the number of messages
+    ///   does not match the number of generators
+    ///
+    /// # Example
+    /// ```rust
+    /// use bbs_mac_vs_sig::{BbsParams, BbsMac};
+    /// use bls12_381::Scalar;
+    /// use ff::Field;
+    ///
+    /// let mut rng = rand::thread_rng();
+    /// let num_messages = 5;
+    /// let params = BbsParams::new(num_messages, &mut rng);
+    /// let x = Scalar::random(&mut rng);
+    /// let mut messages = Vec::with_capacity(num_messages);
+    /// for _ in 0..num_messages {
+    ///     messages.push(Scalar::random(&mut rng));
+    /// }
+    ///
+    /// let mac = BbsMac::sign(&x, &messages, &params, &mut rng).unwrap();
+    /// BbsMac::verify(&x, &messages, &mac, &params).unwrap();
+    /// ```
+    pub fn sign(
```

Comment:
Ah, this is an interesting way of doing the API: basically have a constructor for the BBS MAC that does the actual signing. It's a little unusual though. My taste would be to have `BbsParams` do the signing, as in:

```rust
impl Bbsparams {
  fn sign(&self, messages, ...) -> Result<BbsMac, BbsMacError> {...
}